packages/salesforcedx-vscode-core/src/commands/util/commandLog.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { CommandEvent, CommandEventStream, CommandEventType } from '@salesforce/salesforcedx-utils-vscode';
+import { WorkspaceContext } from '../../context';
+import { CommandLogEntry } from './commandLogEntry';
+
+export const getCommandLog = (commandIdFilter?: string, exitCodeFilter?: number): CommandLogEntry[] => {
+  return CommandLog.getInstance().getCommandLog(commandIdFilter, exitCodeFilter);
+};
+
+export const getLastCommandLogEntry = (commandIdFilter?: string, exitCodeFilter?: number): CommandLogEntry | undefined => {
+  const commandLog = CommandLog.getInstance().getCommandLog(commandIdFilter, exitCodeFilter);
+  return commandLog.length > 0 ? commandLog[commandLog.length - 1] : undefined;
+};
+
+export class CommandLog {
+  private static instance: CommandLog;
+  private static STORAGE_KEY = 'COMMAND_LOG';
+  private static MAX_LOG_ENTRIES = 1000;
+
+  private commandLogEntries: CommandLogEntry[] = [];
+  private inProgressCommands: Map<string, number> = new Map();
+  private inProgressData: any = {};
```

Comment:
Not sure I understand the question completely. But the reason things are set up like this is because of where the code knows what the `commandId` is and where the code does not. The code only knows the `commandId` at command registration. But for other information that can get logged (errors, exit codes, misc data), that data is coming from implementation code where we don't have the `commandId`. The `inProgressCommands` keeps track for commands for which a `START` event has been issued but no `END` yet. The `inProgressData` keeps track of any `EXIT_CODE`, `DATA, `ERROR` events that are issued in between those events.

If the log simply logged the raw events, it would be harder to pull the data like `EXIT_CODE` and associate it with a given command.